src/librustc_mir/lints.rs

```diff
@@ -0,0 +1,132 @@
+
+    // Walk through this function (say `f`) looking to see if
+    // every possible path references itself, i.e. the function is
+    // called recursively unconditionally. This is done by trying
+    // to find a path from the entry node to the exit node that
+    // *doesn't* call `f` by traversing from the entry while
+    // pretending that calls of `f` are sinks (i.e. ignoring any
+    // exit edges from them).
+    //
+    // NB. this has an edge case with non-returning statements,
+    // like `loop {}` or `panic!()`: control flow never reaches
+    // the exit node through these, so one can have a function
+    // that never actually calls itself but is still picked up by
+    // this lint:
+    //
+    //     fn f(cond: bool) {
+    //         if !cond { panic!() } // could come from `assert!(cond)`
+    //         f(false)
+    //     }
+    //
+    // In general, functions of that form may be able to call
+    // itself a finite number of times and then diverge. The lint
+    // considers this to be an error for two reasons, (a) it is
+    // easier to implement, and (b) it seems rare to actually want
+    // to have behaviour like the above, rather than
+    // e.g. accidentally recursing after an assert.
+
+    let basic_blocks = mir.basic_blocks();
+    let mut reachable_without_self_call_queue = vec![mir::START_BLOCK];
+    let mut reached_exit_without_self_call = false;
+    let mut self_call_locations = vec![];
+    let mut visited = BitSet::new_empty(basic_blocks.len());
+
+    let param_env = tcx.param_env(def_id);
+    let caller_substs = Substs::identity_for_item(tcx, def_id);
+
+    while let Some(bb) = reachable_without_self_call_queue.pop() {
+        if visited.contains(bb) {
+            //already done
+            continue;
+        }
+
+        visited.insert(bb);
+
+        let block = &basic_blocks[bb];
+
+        if let Some(ref terminator) = block.terminator {
+            match terminator.kind {
+                TerminatorKind::Call { ref func, .. } => {
+                    let func_ty = func.ty(mir, tcx);
+
+                    if let TyKind::FnDef(fn_def_id, substs) = func_ty.sty {
+                        let (call_fn_id, call_substs) =
+                            if let Some(instance) = Instance::resolve(tcx,
+                                                                        param_env,
+                                                                        fn_def_id,
+                                                                        substs) {
+                                (instance.def_id(), instance.substs)
+                            } else {
+                                (fn_def_id, substs)
+                            };
+
+                        //is this a self call?
+                        let is_self_call =
+                            match tcx.opt_associated_item(def_id) {
+                                Some(assoc_item) if assoc_item.kind == AssociatedKind::Method => {
```

Comment:
The `Method` check is unnecessary, but you need to check whether `assoc_item.container` is a trait.